added shoutout command that returns a tallycoin invoice

Egge7 opened the pull request:
I added a new command that returns a tallycoin lightning invoice. Users can specify an amount and a memo, that will be displayed on tallyco.in. As tallyco.in allows memos up to 140 characters, memos longer than that will be shortened (users wont get notified).

Invoices are generated by tallycoin via their [API ](https://docs.tallyco.in/api/). There is currently no error handling for API failure.

It would be nice if we found a way to inform users about character limit and the podcast specific 21.000 sat shoutout threshold.

Egge7 commented:
I added chat.type as a condition to the shoutout function. If users use /shoutout in a group they will be told to use it in a private chat with the bot instead. I though this might be a good idea to avoid confusion in bigger groups.

I'll modify the return to show qr code + invoice string as soon as i figure out how to do that.

Egge7 commented:
I figured it out and added the function. "/shoutout <amount> <memo>" will now request a lightning invoice from tallyco.in API and return it as QR code + string.

PatrickLemke commented on src/bot.py:
```diff
@@ -91,6 +92,12 @@ def episode_command(update: Update, context: CallbackContext):
 
+def shoutout_command(update: Update, context: CallbackContext):
+    """
+    Returns a TallyCoin LN invoice for a specific amount thats includes a memo
```
```suggestion
    Returns a TallyCoin LN invoice for a specific amount that includes a memo
```

PatrickLemke commented on src/einundzwanzig.py:
```diff
@@ -45,4 +48,39 @@ def episode(update: Update, context: CallbackContext):
+    context.bot.send_message(chat_id=update.effective_chat.id, text=message)
+
+def getInvoice(amt, shoutout):
+    config.TALLYDATA['satoshi_amount'] = str(amt)
```
Using the same object for each request will run into concurrency issues if two users want to send a shoutout at the same time, since the commands are running asynchronously (unlikely that this is going to happen but we should make sure to avoid that problem). You can either create a new object each time the request is sent or clone the existing one.

PatrickLemke commented on src/einundzwanzig.py:
```diff
@@ -45,4 +48,39 @@ def episode(update: Update, context: CallbackContext):
+
+def createQR(text):
+    img = qrcode.make(text)
+    img.save('qr.png')
```
Is it necessary to save that image or can we keep that object in memory and send it from there? We also need to be careful about concurrency issues here.

PatrickLemke reviewed (commented):
I think it makes more sense to use a conversation handler for this feature. The steps would be to send /shoutout to the bot. The bot then responds with a message indicating that you should send the amount next. If that is completed, ask for the message. During the final step we can show the shoutout and message that will be sent and ask the user to confirm with a button. If the button is pressed, display the invoice.

PatrickLemke commented:
> I figured it out and added the function. "/shoutout " will now request a lightning invoice from tallyco.in API and return it as QR code + string.
> 
> Still havn't figured out how to commit as github user though...

I think you have to set the email in your commit to the same email you use for github.

Egge7 commented:
Thanks, that worked. Commits are now linked to the correct user :)

I moved TALLYDATA into the getInvoice function, which should resolve the cuncurrency issue. Also i have changed the createQR function to include chat_id in the file name. The image-files will be removed once they have been sent.

I looked at the telegram convesation handler. However i was not able to figure out how to save values until the end of the conversation. So when the amount is specified in the first part of the convesation, it is easy to pass that value to the second part, but i wasn't able to retrieve this value at the end of the conversation to request an invoice. While this would be very nice to have, i am not sure if this is worth the hassle (unless there is an easy fix for this).

PatrickLemke commented on requirements.txt:
```diff
@@ -1,3 +1,6 @@
-beautifulsoup4
\ No newline at end of file
+beautifulsoup4
+json
```
json and os aren't dependencies that need to be installed, it's part of the standard library. In fact it will lead to an error if you try to install it.

PatrickLemke commented on src/einundzwanzig.py:
```diff
@@ -0,0 +1,99 @@
+    'id': 'zfxqtu',
+    'satoshi_amount': '21',
+    'payment_method': 'ln',
+    'message' : ''
```
I don't think this parameter exists on the tallycoin api? Can we even send a memo?

PatrickLemke commented on src/einundzwanzig.py:
```diff
@@ -0,0 +1,99 @@
+            context.bot.send_message(chat_id=update.effective_chat.id, text= f'Bitte gib einen gültigen Betrag ein')
+    else:
+        context.bot.send_message(chat_id=update.effective_chat.id, text= f'''
+        Shoutouts können nur im direkten Chat mit dem Community Bot gesendet werden. Bitte beginne eine Konvesation mit @einundzwanzigbot!''')
```
```suggestion
        Shoutouts können nur im direkten Chat mit dem Community Bot gesendet werden. Bitte beginne eine Konversation mit @einundzwanzigbot!''')
```

PatrickLemke reviewed (commented):
I think we need to figure out if this feature is even possible given that I don't think the Tallycoin api allows us to send a memo at the moment.

Egge7 commented:
I think i figured it out. However, as my testing envoirment is a sinmgle file, I am not sure if conversations states are imported correctly into bot.py.

/shoutout (only in private chats) will start a conversation with the bot. First the user has to specify a donation amount, then the donation memo. The bot will return a lightninhg invoice + qr code.

PatrickLemke commented on src/einundzwanzig.py:
```diff
@@ -0,0 +1,122 @@
+def invoice(update: Update, context: CallbackContext) -> int:
+    """Stores the message and returns the lightning invoice + qr code."""
+    message = update.message.text
+    context.user_data['message'] = message
```
`user_data` can also be `None` here

PatrickLemke reviewed (commented):
I think we need some fixes still, but if you fix the bugs I pointed out it already works really well. Have a look at my PR, I think we can integrate some more changes from there.

PatrickLemke commented on src/einundzwanzig.py:
```diff
@@ -0,0 +1,122 @@
+    img = qrcode.make(text)
+    img.save(str(chatid) + '.png')
+
+SHOUTOUT_AMOUNT, SHOUTOUT_MEMO = range(2)
```
I would move those to the begging of the file. Hard to find otherwise.

PatrickLemke commented on src/bot.py:
```diff
@@ -113,6 +140,14 @@ def run(bot_token: str):
+    shoutout_handler = ConversationHandler(
+        entry_points=[CommandHandler('shoutout', shoutout_command, run_async=True)],
+        states={
+            SHOUTOUT_AMOUNT: [MessageHandler(Filters.regex('^[0-9]*$'), memo_command, run_async=True)],
```
`SHOUTOUT_AMOUNT` and `SHOUTOUT_MEMO` need to be imported from einundzwanzig.py

Egge7 commented:
I have implemented the fixes suggested in your commit/PR with one exception.
In https://github.com/Egge7/einundzwanzigbot/pull/1 you suggested to return an error, whenever users try to donate less than 21k satoshis. I don't think this is necessary. With my current commit, the bot will point out, that only shoutouts above 21k will be read out in the next podcast episode, however it will still accept donations below 21k.

PatrickLemke commented:
I've added some more fixes, I think this feature is ready to be merged now.

Egge7 commented:
Thanks for the review! I definitely have to change up my workflow, to avoid making mistakes in the future.

I think this looks great, however I am not quite sure if memos that are longer than 140 characters are handled the right way now.

Egge7 commented:
I have added another function to the shoutout conversation. I think this feature is pretty much completed now.
checkPayment will check the payment status of the generated invoice every 20 seconds for 10 minutes. Once the payment ist confirmed, users will receive a "Thank you"-message.

PatrickLemke reviewed (changes requested):
I think we should not check the tallycoin API if the payment was completed, because we can run into multiple issues here:

**Denial of Service Attack**
People could create many invoices that are never paid, therefore the bot will send many many requests to the tallycoin API, causing us to get blocked or our IP address to get banned.

**Code responsibility**
Usually you wouldn't want this part of the code that is responsible for the handling of the conversation to also be responsible to do periodic check ups of an API, especially since the code will still be running in the ConversationHandler. It would be better to hand this functionality over to a new class that has its own thread and informs the user that the invoice has been paid. But I think this would take to much time to implement and is not really necessary for this feature anyway (if the invoice was successfully paid the shoutout is probably successful in 99.99% of cases). I think giving the user the link to the API so he can check himself should be enough. If there is a need for this feature we can add it back later.

I would suggest to revert back to the previous commit and I'll merge it then 👍

Egge7 commented:
Understood! I didn't think of this. I reverted the checkPayment commit. Thanks for the detailed feedback!
